### PR TITLE
feat(ironfish): Move transaction verification to mem pool

### PR DESCRIPTION
**Pull request by rohanjadvani**

### CODE REVIEW

File: ironfish/src/consensus/verifier.ts [hunk 1]
```diff
@@ -176,27 +168,12 @@ export class Verifier {
       throw new Error('Transaction cannot deserialize')
     }
```
Comment (NullSoldier) [inline ironfish/src/consensus/verifier.ts:169]: This is a memory leak now, if you look here were not using `returnReference()` if no exception is thrown. We could move this to a finally clause and remove it in the catch()

File: ironfish/src/memPool/memPool.ts [hunk 2]
```diff
@@ -42,7 +42,12 @@ export class MemPool {
   /**
    * Accepts a transaction from the network
    */
-  acceptTransaction(transaction: Transaction): boolean {
+  async acceptTransaction(transaction: Transaction): Promise<boolean> {
+    const { valid } = await this.chain.verifier.verifyTransaction(transaction)
+    if (!valid) {
```
Comment (NullSoldier) [inline ironfish/src/memPool/memPool.ts:47]: We should add a log here about the invalid transaction and the reason

Comment (NullSoldier) [reply ironfish/src/memPool/memPool.ts:47]: We should add ban score for peers here later, I just haven't thought through all the details yet.

File: ironfish/src/consensus/verifier.ts (hunk 1 above)
Comment (rohanjadvani) [reply ironfish/src/consensus/verifier.ts:170]: Fixed

File: ironfish/src/memPool/memPool.ts [hunk 3]
```diff
@@ -42,15 +42,24 @@ export class MemPool {
...
     const hash = transaction.transactionHash()
+    const renderedHash = hash.toString('hex')
+    if (!valid) {
+      if (reason) {
+        this.logger.debug(`Invalid transaction '${renderedHash}': ${reason}`)
+      }
+      return false
+    }
+
     if (this.transactions.has(hash)) {
```
Comment (NullSoldier) [inline ironfish/src/memPool/memPool.ts:56]: We should re-order these checks

**Review by NullSoldier — APPROVED** [review]

**Pull request merged**